File: C++/2-Medium/00328._奇偶链表/00328._奇偶链表.cpp
#include <iostream>
#include <vector> // 包含 vector 头文件
using namespace std;

// 定义链表节点结构体
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// 实现奇偶链表函数
ListNode* oddEvenList(ListNode* head) {
    if (head == NULL || head->next == NULL) return head;
    
    ListNode* odd = head; // 奇数节点起始点
    ListNode* even = head->next; // 偶数节点起始点
    ListNode* evenHead = even; // 保存偶数链表的头节点
    
    // 遍历链表，重新组织奇偶节点
    while (even != NULL && even->next != NULL) {
        odd->next = even->next; // 将奇数节点指向下一个奇数节点
        odd = odd->next; // 移动奇数节点指针
        even->next = odd->next; // 将偶数节点指向下一个偶数节点
        even = even->next; // 移动偶数节点指针
    }
    odd->next = evenHead; // 连接奇数链表和偶数链表
    
    return head;
}

// 辅助函数，用于创建链表
ListNode* createList(const vector<int>& vals) {
    ListNode dummy(0);
    ListNode* tail = &dummy;
    for (int val : vals) {
        tail->next = new ListNode(val);
        tail = tail->next;
    }
    return dummy.next;
}

// 辅助函数，用于打印链表
void printList(ListNode* head) {
    while (head != NULL) {
        cout << head->val << " -> ";
        head = head->next;
    }
    cout << "NULL" << endl;
}

int main() {
    // 示例链表：1->2->3->4->5->NULL
    vector<int> vals = {1, 2, 3, 4, 5};
    ListNode* head = createList(vals);

    cout << "Original List: ";
    printList(head);

    ListNode* result = oddEvenList(head);

    cout << "Odd-Even List: ";
    printList(result);

    return 0;
}

File: Python/2-Medium/00328._奇偶链表/00328._奇偶链表.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def oddEvenList(head: Optional[ListNode]) -> Optional[ListNode]:
    """
    根据给定的单链表，重新排列以使所有奇数位置的节点位于偶数位置节点之前。
    
    参数:
    head: ListNode - 链表的头节点。
    
    返回:
    ListNode - 重排后的链表的头节点。
    
    示例:
    输入: 1->2->3->4->5->NULL
    输出: 1->3->5->2->4->NULL
    """
    # 边界条件：如果链表为空或只有一个节点，直接返回
    if not head or not head.next:
        return head
    
    # 初始化奇数、偶数指针和偶数头部指针
    odd = head
    even = head.next
    evenHead = head.next
    
    # 遍历链表，重新链接奇偶节点
    while even and even.next:
        odd.next = odd.next.next
        even.next = even.next.next
        # 注意：上下文顺序
        odd = odd.next
        even = even.next
    
    # 将偶数链表拼接到奇数链表后
    odd.next = evenHead
    
    return head

# 辅助函数：创建链表
def create_list(nodes):
    head = ListNode(nodes[0])
    current = head
    for value in nodes[1:]:
        current.next = ListNode(value)
        current = current.next
    return head

# 辅助函数：打印链表
def print_list(head):
    current = head
    while current:
        print(current.val, end=' -> ' if current.next else '')
        current = current.next
    print('NULL')

# 示例
nodes = [1, 2, 3, 4, 5]
head = create_list(nodes)
print("原始链表:")
print_list(head)

# 运行代码
new_head = oddEvenList(head)
print("重排后的链表:")
print_list(new_head)